src/Compilers/CSharp/Portable/Symbols/TypeWithAnnotations.cs

```diff
@@ -50,13 +50,20 @@ internal struct Builder
             /// </remarks>
             internal bool InterlockedInitialize(TypeWithAnnotations type)
             {
-                if ((object)_defaultType != null)
+                if (!IsDefault)
                 {
                     return false;
                 }
                 _nullableAnnotation = type.NullableAnnotation;
-                Interlocked.CompareExchange(ref _extensions, type._extensions, null);
-                return (object)Interlocked.CompareExchange(ref _defaultType, type.DefaultType, null) == null;
+                Interlocked.CompareExchange(ref _defaultType, type.DefaultType, null);
+                return Interlocked.CompareExchange(ref _extensions, type._extensions ?? Extensions.Default, null) is null;
+            }
+
+            internal void InterlockedReset()
```

Comment:
I don't think this method can be safely implemented. I think we need to change the code pattern that requires it's uses.

Reply:
It's also completely broken from the general symbol model -- Symbols can never change their type.

Reply:
This is only done in a symbol constructor, ie before any other thread could access the fields.

Reply:
I looked into this more following our discussion. Trying to fix the existing problem with `SourcePropertySymbol` (which can set its type a second time, mid-construction) is a bit tricky. Also, I do not think it's necessarily worth it because this only happens in an error scenario.

For convenience, I captured the tests that are affected. Most are just producing additional diagnostics (which is not bad, and we could probably improve), but some also result in different custom modifiers in symbol APIs. This means that we're indeed reading the first type that is set and making decisions off of it before changing our minds.

We can file a separate issue to follow-up on that. For convenience, I captured the tests that are impacted:

![image](https://user-images.githubusercontent.com/12466233/56998110-e08dd480-6b5e-11e9-82b3-e1dfb76bfd03.png)

Reply:
Filed https://github.com/dotnet/roslyn/issues/35381

Reply:
I'd use a name like `Dangerous` here so we don't get any further uses.

Reply:
And EditorBrowseable(false)